red_black_tree/rb_tree.py

```diff
@@ -0,0 +1,531 @@
+        else:
+            self.root = node
+
+    def _right_rotation(self, node, parent, grandfather, to_recolor=False):
+        grand_grandfather = grandfather.parent
+        self.__update_parent(node=parent, parent_old_child=grandfather, new_parent=grand_grandfather)
+
+        old_right = parent.right
+        parent.right = grandfather
+        grandfather.parent = parent
+
+        grandfather.left = old_right  # save the old right values
+        old_right.parent = grandfather
+
+        if to_recolor:
+            parent.color = BLACK
+            node.color = RED
+            grandfather.color = RED
+
+    def _left_rotation(self, node, parent, grandfather, to_recolor=False):
+        grand_grandfather = grandfather.parent
+        self.__update_parent(node=parent, parent_old_child=grandfather, new_parent=grand_grandfather)
+
+        old_left = parent.left
+        parent.left = grandfather
+        grandfather.parent = parent
+
+        grandfather.right = old_left  # save the old left values
+        old_left.parent = grandfather
+
+        if to_recolor:
+            parent.color = BLACK
+            node.color = RED
+            grandfather.color = RED
+
+    def _recolor(self, grandfather):
+        grandfather.right.color = BLACK
+        grandfather.left.color = BLACK
+        if grandfather != self.root:
+            grandfather.color = RED
+        self._try_rebalance(grandfather)
+
+    def _find_parent(self, value):
+        """ Finds a place for the value in our binary tree"""
+        def inner_find(parent):
+            """
+            Return the appropriate parent node for our new node as well as the side it should be on
+            """
+            if value == parent.value:
+                return None, None
+            elif parent.value < value:
+                if parent.right.color == NIL:  # no more to go
+                    return parent, 'R'
+                return inner_find(parent.right)
+            elif value < parent.value:
+                if parent.left.color == NIL:  # no more to go
+                    return parent, 'L'
+                return inner_find(parent.left)
+
+        return inner_find(self.root)
+
+    def find_node(self, value):
+        def inner_find(root):
+            if root is None or root == self.NIL_LEAF:
+                return None
+            if value > root.value:
+                return inner_find(root.right)
+            elif value < root.value:
+                return inner_find(root.left)
+            else:
+                return root
+
+        found_node = inner_find(self.root)
+        return found_node
+
+    def _find_in_order_successor(self, node):
+        right_node = node.right
+        left_node = right_node.left
+        if left_node == self.NIL_LEAF:
+            return right_node
+        while left_node.left != self.NIL_LEAF:
+            left_node = left_node.left
+        return left_node
+
+    def _get_sibling(self, node):
```

Comment:
I mean yes it can, but its coupled with the Node class. I could make it be a static method of the Node class